src/server/tiered_storage.cc

```diff
@@ -396,9 +404,17 @@ error_code TieredStorage::ScheduleOffload(DbIndex db_index, PrimeIterator it) {
   if (db_arr_[db_index] == nullptr) {
     db_arr_[db_index] = new PerDb;
   }
+  unsigned max_pending_writes = GetFlag(FLAGS_tiered_storage_max_pending_writes);
+  unsigned throttle_usec = GetFlag(FLAGS_tiered_storage_throttle_us);
+  if (num_active_requests_ >= max_pending_writes && throttle_usec > 0) {
+    chrono::steady_clock::time_point next =
+        chrono::steady_clock::now() + chrono::microseconds(throttle_usec);
+    stats_.throttled_write_cnt++;
+    throttle_ec_.await_until([&]() { return num_active_requests_ < max_pending_writes; }, next);
```

Comment:
after you await you must check that the iterator is still valid and points to the value that you wanted to offload

Reply:
also the same check needs to be done in the code that calls to ScheduleOffload

Reply:
I see TODOs now. I prefer to wait for your PR that provides these checks. Meanwhile I will add another TODO